File: Nori/Model/ClipRow.swift
import Foundation

/// Everything a card needs, as a Sendable value. Built by `HistoryStore` from `ClipItem`,
/// by `SensitiveVault` for masked secrets, and by the panel for ghost rows.
struct ClipRow: Identifiable, Hashable, Sendable {
    enum Source: Sendable { case history, sensitive, ghost }

    let id: UUID
    let source: Source
    let kind: ClipKind
    let title: String
    let searchText: String
    let sourceBundleID: String?
    let sourceAppName: String?
    let isFromUniversalClipboard: Bool
    let firstCopiedAt: Date
    let lastCopiedAt: Date
    let copyCount: Int
    let pinnedAt: Date?
    let byteCount: Int
    let characterCount: Int
    let lineCount: Int
    let isRichText: Bool
    let isTruncated: Bool
    let linkURL: URL?
    let colorHex: String?
    let fileURLs: [URL]
    let imagePixelSize: CGSize?
    let thumbnail: Data?
    /// Sensitive rows only: when the in-memory copy disappears.
    let expiresAt: Date?
    /// Ghost rows only: the human-readable reason nothing was saved.
    let ghostReason: String?

    var isPinned: Bool { pinnedAt != nil }
    var isSensitive: Bool { source == .sensitive }
    var isGhost: Bool { source == .ghost }

    /// First non-empty line with runs of whitespace collapsed — what the card's title line shows.
    var displayTitle: String {
        let firstLine = title.split(whereSeparator: \.isNewline).first.map(String.init) ?? title
        return firstLine.split(whereSeparator: \.isWhitespace).joined(separator: " ")
    }

    /// The first two lines, leading whitespace preserved (code cards).
    var codeLines: [String] {
        let expanded = title.replacingOccurrences(of: "\t", with: "    ")
        let lines = expanded.split(omittingEmptySubsequences: false, whereSeparator: \.isNewline).map(String.init)
        let firstNonEmpty = lines.firstIndex { !$0.trimmingCharacters(in: .whitespaces).isEmpty } ?? 0
        return Array(lines[firstNonEmpty...].prefix(2))
    }

    var linkHost: String? {
        guard let host = linkURL?.host() else { return nil }
        return host.hasPrefix("www.") ? String(host.dropFirst(4)) : host
    }

    /// Path + query of a link, or nil when it is just the root.
    var linkPathAndQuery: String? {
        guard let url = linkURL else { return nil }
        var path = url.path(percentEncoded: false)
        if let query = url.query(percentEncoded: false), !query.isEmpty { path += "?" + query }
        return path == "/" || path.isEmpty ? nil : path
    }

    static func ghost(reason: String, at date: Date) -> ClipRow {
        ClipRow(
            id: UUID(), source: .ghost, kind: .text, title: reason, searchText: "", sourceBundleID: nil,
            sourceAppName: nil, isFromUniversalClipboard: false, firstCopiedAt: date, lastCopiedAt: date,
            copyCount: 0, pinnedAt: nil, byteCount: 0, characterCount: 0, lineCount: 0, isRichText: false,
            isTruncated: false, linkURL: nil, colorHex: nil, fileURLs: [], imagePixelSize: nil, thumbnail: nil,
            expiresAt: nil, ghostReason: reason
        )
    }
}

File: windows/Nori.Core/ClipRow.cs
namespace Nori.Core;

public enum RowSource
{
    History,
    Sensitive,
    Ghost,
}

/// <summary>
/// Everything a card needs, as an immutable value. Built by <see cref="HistoryStore"/> from stored
/// items, by <see cref="SensitiveVault"/> for masked secrets, and by the panel for ghost rows.
/// </summary>
public sealed record ClipRow
{
    public required Guid Id { get; init; }
    public required RowSource Source { get; init; }
    public required ClipKind Kind { get; init; }
    public required string Title { get; init; }
    public string SearchText { get; init; } = string.Empty;
    public string? SourceApp { get; init; }
    public string? SourceAppName { get; init; }
    public string? SourceAppPath { get; init; }
    public required DateTimeOffset FirstCopiedAt { get; init; }
    public required DateTimeOffset LastCopiedAt { get; init; }
    public int CopyCount { get; init; } = 1;
    public DateTimeOffset? PinnedAt { get; init; }
    public int ByteCount { get; init; }
    public int CharacterCount { get; init; }
    public int LineCount { get; init; }
    public bool IsRichText { get; init; }
    public bool IsTruncated { get; init; }
    public Uri? LinkUrl { get; init; }
    public string? ColorHex { get; init; }
    public IReadOnlyList<string> FilePaths { get; init; } = [];
    public (int Width, int Height)? ImagePixelSize { get; init; }
    public byte[]? Thumbnail { get; init; }

    /// <summary>Sensitive rows only: when the in-memory copy disappears.</summary>
    public DateTimeOffset? ExpiresAt { get; init; }

    /// <summary>Ghost rows only: why nothing was saved.</summary>
    public GhostReason? GhostReason { get; init; }

    public bool IsPinned => PinnedAt is not null;
    public bool IsSensitive => Source == RowSource.Sensitive;
    public bool IsGhost => Source == RowSource.Ghost;

    /// <summary>First non-empty line with runs of whitespace collapsed — what the card's title line shows.</summary>
    public string DisplayTitle
    {
        get
        {
            var firstLine = KindDetector.SplitLines(Title).FirstOrDefault(l => l.Trim().Length > 0) ?? Title;
            return string.Join(' ', firstLine.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
        }
    }

    /// <summary>The first two lines, leading whitespace preserved (code cards).</summary>
    public IReadOnlyList<string> CodeLines
    {
        get
        {
            var expanded = Title.Replace("\t", "    ");
            var lines = KindDetector.SplitLines(expanded);
            var firstNonEmpty = lines.FindIndex(l => l.Trim(' ').Length > 0);
            if (firstNonEmpty < 0) firstNonEmpty = 0;
            return lines.Skip(firstNonEmpty).Take(2).ToList();
        }
    }

    /// <summary>mailto: links have no host to show (.NET parses the address domain as Host).</summary>
    public bool IsMailto => LinkUrl is not null && LinkUrl.Scheme.Equals("mailto", StringComparison.OrdinalIgnoreCase);

    public string? LinkHost
    {
        get
        {
            if (LinkUrl is null || IsMailto || string.IsNullOrEmpty(LinkUrl.Host)) return null;
            var host = LinkUrl.Host;
            return host.StartsWith("www.", StringComparison.OrdinalIgnoreCase) ? host[4..] : host;
        }
    }

    /// <summary>Path + query of a link, or null when it is just the root.</summary>
    public string? LinkPathAndQuery
    {
        get
        {
            if (LinkUrl is null || IsMailto || !LinkUrl.IsAbsoluteUri || string.IsNullOrEmpty(LinkUrl.Host)) return null;
            var path = Uri.UnescapeDataString(LinkUrl.AbsolutePath);
            var query = LinkUrl.Query;
            if (query.Length > 1) path += Uri.UnescapeDataString(query);
            return path is "/" or "" ? null : path;
        }
    }

    /// <summary>File names for file cards ("3 files" is rendered by the UI when there are several).</summary>
    public IReadOnlyList<string> FileNames => FilePaths.Select(ClipClassifier.FileName).ToList();

    public static ClipRow Ghost(GhostReason reason, DateTimeOffset at, string? appName = null) => new()
    {
        Id = Guid.NewGuid(),
        Source = RowSource.Ghost,
        Kind = ClipKind.Text,
        Title = reason switch
        {
            GhostReason.Concealed c => $"Concealed item from {c.AppName ?? appName ?? "an app"} wasn't saved",
            GhostReason.ImageTooLarge i => $"Image too large ({ByteSize.Format(i.Bytes)}) wasn't saved",
            _ => "Item wasn't saved",
        },
        FirstCopiedAt = at,
        LastCopiedAt = at,
        CopyCount = 0,
        GhostReason = reason,
    };
}

/// <summary>"412 KB" style sizes, matching the card captions.</summary>
public static class ByteSize
{
    public static string Format(long bytes)
    {
        if (bytes < 1000) return $"{bytes} B";
        if (bytes < 1000 * 1000) return $"{bytes / 1000.0:0} KB";
        if (bytes < 1000L * 1000 * 1000) return $"{bytes / 1_000_000.0:0.#} MB";
        return $"{bytes / 1_000_000_000.0:0.##} GB";
    }
}
